### PR TITLE
Add non-deterministic warning

### PR DESCRIPTION
**Pull request by Crissman**
To address issue #4134.

Created @non_deterministic decorator.
Added decorator to functions that use atomicAdd

### CODE REVIEW

**Comment by niboshi** [comment]
I think it's better to implement it with a function rather than a decorator, because in general determinism depends on runtime information. I think there's no advantage of a decorator over a function.

By the way, shouldn't we also annotate nondeterminism in numpy?

**Comment by Crissman** [comment]
Actually, I started replacing the decorator with a function, and I don't think it's better. Grammatically, it's cleaner to have a `@non_deterministic` decorator above the relevant functions. If instead of the decorator, I added a function call within the function, what would I call it? `check_deterministic()`? If so, that feels confusing with `config.check_deterministic`. I think it's clearer to use the decorator, since it makes for a cleaner name space.

**Comment by niboshi** [comment]
First of all, I don't think `config.check_deterministic` is good. Its behavior is ambiguous. After checking determinism, what does it do? Keep in a log? Raise warning? Does it warn if it's deterministic? Or opposite?

I would prefer something like `config.warn_nondeterministic`.

As for the function name, how about simply `util.nondeterministic()`?
The role of this function would be "to declare the execution path is nondeterministic".

**Comment by Crissman** [comment]
@niboshi I'm convinced. Converted to a function instead of a decorator.

**Comment by Crissman** [comment]
Travis error on Mac appears unrelated.

**Comment by niboshi** [comment]
We won't be able to assure 100% detection rate. We should mention that this feature is provided as best-effort and users should not rely on it.
I think we should also mention the possibilities of false-positives: for example, reduction on a single element is deterministic but I don't think we will eliminate every such conditions.

File: chainer/functions/connection/embed_id.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 from chainer.backends import cuda
 from chainer import function_node
 from chainer.utils import type_check
+from chainer.utils import non_deterministic
```
Comment (niboshi) [inline chainer/functions/connection/embed_id.py]: "Do not import objects, only modules".
https://docs.openstack.org/hacking/latest/user/hacking.html#imports

File: chainer/utils/non_deterministic.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import warnings
+from chainer import configuration
+
+
+def non_deterministic(f_name):
```
Comment (niboshi) [inline chainer/utils/non_deterministic.py]: How about `nondeterministic`?

**Comment by Crissman** [comment]
@niboshi Please review revisions.

File: chainer/functions/connection/embed_id.py [hunk 3]
```diff
@@ -59,6 +60,7 @@ def __init__(self, w_shape, ignore_label=None):
         self.ignore_label = ignore_label
 
     def forward(self, inputs):
+        utils.nondeterministic('atomicAdd')
```
Comment (niboshi) [inline chainer/functions/connection/embed_id.py]: It seems this should be placed in the cupy code path.

File: chainer/functions/loss/ctc.py [hunk 4]
```diff
@@ -154,6 +154,7 @@ def log_matrix(self, x, xp):
     # path probability to label probability
     def label_probability(self, label_size, path, path_length,
                           multiply_seq, xp):
+        utils.nondeterministic('atomicAdd')
```
Comment (niboshi) [inline chainer/functions/loss/ctc.py]: ditto.

File: docs/source/reference/configuration.rst [hunk 5]
```diff
@@ -66,6 +66,11 @@ Configuration Keys
 
    You can change the default value to ``True`` by setting ``CHAINER_KEEP_GRAPH_ON_REPORT`` environment variable to ``1``.
 
+* ``warn_nondeterministic`` (default: ``False``)
+   Flag to give warning when a non-deterministic function is used. This function is experimental.
+
+   If it is true, then functions that use non-deterministic functions, such as atomicAdd, will give a warning when executed. Note the determinisitic outputs may still result, despite using a nondeterministic function, depending on usage.
```
Comment (niboshi) [inline docs/source/reference/configuration.rst]: We should mention false negatives first, which are more important.
Let's discuss the expression offline later.

File: chainer/utils/nondeterministic.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import warnings
+from chainer import configuration
```
Comment (niboshi) [inline chainer/utils/nondeterministic.py]: Please add a blank line before importing chainer module.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by Crissman** [comment]
Jenkins error looks unrelated. Retest?

**Comment by niboshi** [comment]
Discussed:
We may raise warning for features that use random numbers by the following additions.
* Introduce an interface to set random seeds
  * Make it possible to enumerate backends (numpy, cupy, chainerx) and set random seeds in a unified API.
* Raise warning in features that use random numbers only if the seed is not set via the above interface.

However we don't do that for now, and make it users' responsibility to set seeds.

@Crissman Let's document that it won't raise warning for such features.

**Comment by Crissman** [comment]
Documentation added that setting seed is responsibility of user.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by Crissman** [comment]
Jenkins issue looks unrelated. Please merge.

**Comment by niboshi** [comment]
@Crissman 
It seems you did the merge wrong. Please fix it.

**Comment by Crissman** [comment]
Ok. I think it's clean now. Please take a look.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
LGTM

**Pull request merged**